news/blog/2022: MGR TTL Cache blog post (#332)
ceph/ceph.io: src/en/news/blog/2022/mgr-ttlcache/index.md

---
title: "Caching mgr modules"
date: "2022-01-19"
author: "Pere Diaz Bou"
image: "images/python.jpg"
tags:
  - "mgr"
  - "performance"
  - "python"
---

## Introduction

> For a time there has been arising problems regarding the scalability of the mgr daemon and the underlying tightly
> coupled (to the daemon) structure of python modules that we know as: dashboard, prometheus, pg_autoscaler, orchestrators and many more.

Reviewer: ```suggestion
Problems have been arising a while ago related to the scalability of the mgr daemon and the underlying tightly
coupled (to the daemon) structure of python modules like dashboard, prometheus, pg_autoscaler, orchestrators and many more.
```

Cluster data is requested on demand by the modules every certain time in order to keep the cluster
status updated in these modules. For example, the dashboard will keep requesting health data updates
to the daemon periodically to display if the cluster is healthy.

There are instances of data structures where scalability is a problem. Osds for example can increase up to thousands, therefore,

> structures like OSDMap will increase proportionally in size and thus, increasing the latency of modules when they try to reach for

Reviewer: ```suggestion
structures like OSDMap will increase proportionally in size and thus, increasing the latency of modules when they retrieve
```

the structure to json (With JsonFormatter) where we only have to worry about serializing and deserializing.

With json serialization in place, a shared TTL Cache between modules and the daemon was added:

![MGR daemon cached](images/mgrarch2.png)

Not everything is cached as it might seem in the diagram but most of the problematic calls are. Serializing with JSON itself
is not a great improvement but, if we cache the serialized object we see **2.8x improvement with 3000 osds**.

In the following line chart a clear distiction is made between the current behaviour and the cache with json serialization. This
benchmarks were ran by requesting the OSDMap 5000 times for each number of osds. These are the results:

![TTL Cache](images/chart.png)

The blue dashed line represents the improvement in latency to retrieve and osdmap, the black line is the usual behaviour and the blue line,
the results without caching json objects.

Moreover, the cache is available in master/quincy and soon to be backported to pacific. It can be enabled by running:

`ceph config set mgr mgr ttl_cache_expire_seconds 10`

> That will set a 10 seconds _time to live_ to each object cached.

Reviewer: ```suggestion
That will set a 10 seconds _time to live_ to each cached object.
```

## Conclusion

With a disabled cache the usual behaviour is still faster, nevertheless, we see a clear improvement of almost 3x on larger clusters with

> the cache enabled and using a simple serialization tool like json is.

Reviewer: ```suggestion
the cache enabled and using a simple serialization tool like json.
```

The mgr daemon has been unattended for a while and a lot of improvements are needed to have a more scalable mgr. This cache is an improvement

> that alleviates some of the workload, but there is more work to be done addresing scalability, debugabilitty, subinterpeters...

Reviewer: ```suggestion
that alleviates some of the workload, but there is more work to be done addressing scalability, debuggabilitty, subinterpreters...
```
Also you can end with: `Stay Tuned!` :smile:
